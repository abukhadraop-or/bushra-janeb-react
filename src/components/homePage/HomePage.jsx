import { Article } from "components/Articles/Articles";
import Banner from "components/Banner/Banner";
import { Pagination } from "components/Pagination/Pagination";
import Tag from "components/Tags/Tags";
import React, { useEffect, useState } from "react";
import { getArticles } from "services/fake-articles-service";
import { getTags } from "services/fake-tags-service";
import paginate from "utils/paginate";
import HomePageContainer from "./home-page.styles";

/**
 * Render the <HomePage> component.
 *
 * @return {JSX.Element}
 */
export const Home = () => {
  const [articles, setArticles] = useState([]);
  const [currentPage, setCurrentPage] = useState(1);
  const [currentTag, setCurrentTag] = useState("");
  const filtered = currentTag
    ? articles.filter((article) => article.title === currentTag.name)
    : articles;
  const [pageSize] = useState(10);
  const filteredArticles = paginate(filtered, currentPage, pageSize);
  const [tags, setTags] = useState([]);

  // Handles initial state.
  useEffect(() => {
    const getData = async () => {
      const result = await getArticles();
      const tag = await getTags();
      setArticles(result);
      setTags(tag);
    };

    getData();
  }, []);

  /**
   * Handle onPageChange event.
   *
   * @param {Number} page The number of the current page.
   */
  const handlePageChange = (page) => {
    setCurrentPage(page);
    window.scrollTo({
      top: 0,
      left: 100,
      behavior: "smooth",
    });
  };

  /**
   * Handle tagSelect event.
   *
   * @param {Object} tag The selected tag.
   */
  const handleTagSelect = (tag) => {
    setCurrentTag(tag);
    setCurrentPage(1);
  };
  return (
    <>
      <Banner />
      <HomePageContainer>
        <Article filteredArticles={filteredArticles} />
        <Tag
          onTagSelect={handleTagSelect}
          selectedTag={currentTag}
          tags={tags}
        />
      </HomePageContainer>
      <Pagination
        currentPage={currentPage}
        itemCount={filtered.length}
        onPageChange={handlePageChange}
        pageSize={pageSize}
      />
    </>
  );
};

export default Home;
