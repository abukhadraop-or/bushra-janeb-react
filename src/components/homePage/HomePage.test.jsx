import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Enzyme, { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { Home } from './HomePage';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('../../services/fake-articles-service');
jest.mock('../../services/fake-tags-service');

// const mockedArticles = [
//   {
//     id: 1,
//     name: 'Gerome',
//     publishDate: 'October 8, 2021',
//     title: 'ReactJs',
//     body: 'Exemplary fullstack Medium.com clone powered by React, Angular, Node, Django, and many more',
//     imgUrl:
//       'https://bombaymeatco.com/wp-content/uploads/2014/11/free-profile-photo-whatsapp-4.png',
//     description:
//       'See how the exact same Medium.com clone (called Conduit) is built using different frontends and backends. Yes, you can mix and match them, because they all adhere to the same API spec',
//   },
// ];

// const mockedTags = [
//   {
//     id: 1,
//     name: 'C++',
//   },
// ];

describe('<Home />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Home />);
  });

  it('should match snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  // it('should send the filtered articles to the article component', async () => {
  //   const wrapper = shallow(<Home />);
  //   const promise = Promise.resolve();
  //   getArticles.mockResolvedValue(mockedArticles);
  //   // const result = await getArticles();
  //   // expect(result).toEqual(mockedArticles);

  //   await act(() => promise);
  //   wrapper.update();
  //   const articleCard = wrapper.find(ArticleCard);
  //   expect(articleCard.props()).toEqual(mockedArticles);
  // });

  // it('should send the tags to the tags component', async () => {
  //   const promise = Promise.resolve();
  //   getTags.mockResolvedValue(mockedTags);

  //   await act(() => promise);
  //   wrapper.update();
  //   const tags = wrapper.find(<Tag />).at(3);
  //   expect(tags).toEqual(mockedTags);
  // });
});
