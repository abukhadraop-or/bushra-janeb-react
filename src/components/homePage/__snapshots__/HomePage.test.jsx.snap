// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should match snapshot 1`] = `
<Fragment>
  <Banner />
  <styled.div>
    <styled.div>
      <Article />
      <ArticleCard
        filteredArticles={Array []}
      />
    </styled.div>
    <Tag
      onTagSelect={[Function]}
      selectedTag=""
      tags={Array []}
    />
  </styled.div>
  <Pagination
    currentPage={1}
    itemCount={0}
    onPageChange={[Function]}
    pageSize={10}
  />
</Fragment>
`;

exports[`Home should match snapshot 1`] = `
<Fragment>
  <Banner />
  <styled.div>
    <styled.div>
      <Article />
      <ArticleCard
        filteredArticles={Array []}
      />
    </styled.div>
    <Tag
      onTagSelect={[Function]}
      selectedTag=""
      tags={Array []}
    />
  </styled.div>
  <Pagination
    currentPage={1}
    itemCount={0}
    onPageChange={[Function]}
    pageSize={10}
  />
</Fragment>
`;
